File: app/test/live-events.test.mjs
// CANLI OLAY ŞERİDİ + BASKI GÖSTERGESİ TESTLERİ (saf modül).
// Kural: uydurma dakika/uydurma baskı yok; veri yetersizse null/boş döner.
import test from 'node:test';
import assert from 'node:assert/strict';
import {
  normalizeEvent, normalizeEvents, timelineMarkers, positionOf,
  goalProgression, pressureIndex, sortStats, REG_MINUTES,
} from '../src/liveEvents.js';

const EVENTS = [
  { minute: 23, type: 'Goal', detail: 'Normal Goal', side: 'home', player: 'A' },
  { minute: 41, type: 'Card', detail: 'Yellow Card', side: 'away', player: 'B' },
  { minute: 12, type: 'subst', detail: 'Substitution 1', side: 'home', player: 'C' },
  { minute: 67, type: 'Card', detail: 'Red Card', side: 'away', player: 'D' },
  { minute: 90, extra: 3, type: 'Goal', detail: 'Penalty', side: 'away', player: 'E' },
];

test('olaylar: normalize + dakikaya göre sıralanır', () => {
  const evs = normalizeEvents(EVENTS);
  assert.deepEqual(evs.map((e) => e.at), [12, 23, 41, 67, 93]);
  assert.equal(evs[1].kind, 'goal');
  assert.equal(evs[2].kind, 'yellow');
  assert.equal(evs[3].kind, 'red');
  assert.equal(evs[4].penalty, true);
});

test('olaylar: dakikasız / tanınmayan olay ATILIR (uydurma dakika yok)', () => {
  assert.equal(normalizeEvent({ type: 'Goal', side: 'home' }), null, 'dakika yoksa olay yok');
  assert.equal(normalizeEvent({ minute: 10, type: 'Whatever' }), null, 'bilinmeyen tür atılır');
  assert.equal(normalizeEvent(null), null);
  assert.deepEqual(normalizeEvents(null), []);
});

test('olaylar: VAR ile iptal edilen gol GOL SAYILMAZ (resmî skorla çelişmez)', () => {
  const evs = normalizeEvents([{ minute: 30, type: 'Goal', detail: 'Goal cancelled', side: 'home' }]);
  assert.deepEqual(evs, []);
  assert.deepEqual(goalProgression([{ minute: 30, type: 'Goal', detail: 'Goal Disallowed', side: 'home' }]), []);
});

test('şerit: yalnız gol ve kırmızı kart gösterilir', () => {
  const marks = timelineMarkers(EVENTS);
  assert.deepEqual(marks.map((m) => m.kind), ['goal', 'red', 'goal']);
  assert.ok(marks.every((m) => m.pos >= 0 && m.pos <= 1), 'konum 0..1 arasında');
  assert.deepEqual(timelineMarkers([]), [], 'olay yoksa şerit yok');
});

test('şerit: uzatma dakikası taşmaz, 90+ sona oturur', () => {
  assert.equal(positionOf(45, 0), 0.5);
  assert.equal(positionOf(0, 0), 0);
  assert.equal(positionOf(120, 0), 1, 'kap dışına taşmaz');
  const last = timelineMarkers(EVENTS).at(-1);
  assert.ok(last.pos > 0.9 && last.pos <= 1, "90+3' sona yakın");
});

test('gol akışı: koşan skor doğru, kendi kalesine karşı takıma yazılır', () => {
  const prog = goalProgression([
    { minute: 10, type: 'Goal', detail: 'Normal Goal', side: 'home' },
    { minute: 20, type: 'Goal', detail: 'Own Goal', side: 'home' },
    { minute: 30, type: 'Goal', detail: 'Penalty', side: 'away' },
  ]);
  assert.deepEqual(prog.map((g) => `${g.home}-${g.away}`), ['1-0', '1-1', '1-2']);
  assert.equal(prog[1].ownGoal, true);
  assert.equal(prog[1].side, 'away', 'kendi kalesine gol rakibe yazılır');
  assert.equal(prog[2].penalty, true);
});

test('gol akışı: takımı belirsiz gol skora yazılmaz', () => {
  assert.deepEqual(goalProgression([{ minute: 5, type: 'Goal', detail: 'Normal Goal' }]), []);
});

const STATS = [
  { type: 'Ball Possession', home: '58%', away: '42%' },
  { type: 'Total Shots', home: 12, away: 6 },
  { type: 'Shots on Goal', home: 5, away: 1 },
  { type: 'Fouls', home: 8, away: 11 },
];

test('baskı: gerçek istatistiklerden pay hesaplanır, toplam 100', () => {
  const pr = pressureIndex(STATS);
  assert.ok(pr, 'yeterli veri var');
  assert.equal(pr.home + pr.away, 100);
  assert.ok(pr.home > 60, 'her ölçüde üstün takım baskıda görünür');
  assert.ok(pr.basis.includes('Şut') && pr.basis.includes('Topla oynama'));
  assert.ok(!pr.basis.includes('Faul'), 'faul baskı ölçüsü değil');
});

test('baskı: yetersiz veride GÖSTERGE ÜRETİLMEZ (uydurma yok)', () => {
  assert.equal(pressureIndex([]), null);
  assert.equal(pressureIndex(null), null);
  assert.equal(pressureIndex([{ type: 'Total Shots', home: 4, away: 2 }]), null, 'tek ölçü yetmez');
  assert.equal(pressureIndex([{ type: 'Total Shots', home: null, away: 2 }, { type: 'Corner Kicks', home: 3, away: null }]), null, 'eksik taraf sayılmaz');
  assert.equal(pressureIndex([{ type: 'Total Shots', home: 0, away: 0 }, { type: 'Corner Kicks', home: 0, away: 0 }]), null, '0-0 veri taşımaz');
});

test('baskı: tek eksik ölçü diğerlerini bozmaz', () => {
  const pr = pressureIndex([
    { type: 'Total Shots', home: 10, away: 10 },
    { type: 'Corner Kicks', home: 5, away: 5 },
    { type: 'Shots on Goal', home: null, away: 3 },
  ]);
  assert.equal(pr.home, 50);
  assert.equal(pr.basis.length, 2);
});

test('istatistik sıralama: önemli satırlar üstte, bilinmeyen sona', () => {
  const sorted = sortStats([
    { type: 'Offsides' }, { type: 'Total Shots' }, { type: 'Bilinmeyen' }, { type: 'Ball Possession' },
  ]);
  assert.deepEqual(sorted.map((s) => s.type), ['Ball Possession', 'Total Shots', 'Offsides', 'Bilinmeyen']);
  assert.equal(REG_MINUTES, 90);
});

File: test/live_events_test.dart
// KAYNAK: app/test/live-events.test.mjs — BİREBİR çeviri.
//
// CANLI OLAY ŞERİDİ + BASKI GÖSTERGESİ TESTLERİ (saf modül).
// Kural: uydurma dakika/uydurma baskı yok; veri yetersizse null/boş döner.

import 'package:flutter_test/flutter_test.dart';
import 'package:masteranaliz/core/live_events.dart';

const List<Map<String, Object?>> _events = [
  {
    'minute': 23,
    'type': 'Goal',
    'detail': 'Normal Goal',
    'side': 'home',
    'player': 'A',
  },
  {
    'minute': 41,
    'type': 'Card',
    'detail': 'Yellow Card',
    'side': 'away',
    'player': 'B',
  },
  {
    'minute': 12,
    'type': 'subst',
    'detail': 'Substitution 1',
    'side': 'home',
    'player': 'C',
  },
  {
    'minute': 67,
    'type': 'Card',
    'detail': 'Red Card',
    'side': 'away',
    'player': 'D',
  },
  {
    'minute': 90,
    'extra': 3,
    'type': 'Goal',
    'detail': 'Penalty',
    'side': 'away',
    'player': 'E',
  },
];

const List<Map<String, Object?>> _stats = [
  {'type': 'Ball Possession', 'home': '58%', 'away': '42%'},
  {'type': 'Total Shots', 'home': 12, 'away': 6},
  {'type': 'Shots on Goal', 'home': 5, 'away': 1},
  {'type': 'Fouls', 'home': 8, 'away': 11},
];

void main() {
  test('olaylar: normalize + dakikaya göre sıralanır', () {
    final evs = normalizeEvents(_events);
    expect([for (final e in evs) e.at], [12, 23, 41, 67, 93]);
    expect(evs[1].kind, 'goal');
    expect(evs[2].kind, 'yellow');
    expect(evs[3].kind, 'red');
    expect(evs[4].penalty, isTrue);
  });

  test('olaylar: dakikasız / tanınmayan olay ATILIR (uydurma dakika yok)', () {
    expect(
      normalizeEvent({'type': 'Goal', 'side': 'home'}),
      isNull,
      reason: 'dakika yoksa olay yok',
    );
    expect(
      normalizeEvent({'minute': 10, 'type': 'Whatever'}),
      isNull,
      reason: 'bilinmeyen tür atılır',
    );
    expect(normalizeEvent(null), isNull);
    expect(normalizeEvents(null), isEmpty);
  });

  test('olaylar: VAR ile iptal edilen gol GOL SAYILMAZ', () {
    expect(
      normalizeEvents([
        {
          'minute': 30,
          'type': 'Goal',
          'detail': 'Goal cancelled',
          'side': 'home',
        },
      ]),
      isEmpty,
    );
    expect(
      goalProgression([
        {
          'minute': 30,
          'type': 'Goal',
          'detail': 'Goal Disallowed',
          'side': 'home',
        },
      ]),
      isEmpty,
    );
  });

  test('şerit: yalnız gol ve kırmızı kart gösterilir', () {
    final marks = timelineMarkers(_events);
    expect([for (final m in marks) m.e.kind], ['goal', 'red', 'goal']);
    expect(
      marks.every((m) => m.pos >= 0 && m.pos <= 1),
      isTrue,
      reason: 'konum 0..1 arasında',
    );
    expect(timelineMarkers([]), isEmpty, reason: 'olay yoksa şerit yok');
  });

  test('şerit: uzatma dakikası taşmaz, 90+ sona oturur', () {
    expect(positionOf(45, 0), 0.5);
    expect(positionOf(0, 0), 0);
    expect(positionOf(120, 0), 1, reason: 'kap dışına taşmaz');
    final last = timelineMarkers(_events).last;
    expect(last.pos > 0.9 && last.pos <= 1, isTrue, reason: "90+3' sona yakın");
  });

  test('gol akışı: koşan skor doğru, kendi kalesine karşı takıma yazılır', () {
    final prog = goalProgression([
      {'minute': 10, 'type': 'Goal', 'detail': 'Normal Goal', 'side': 'home'},
      {'minute': 20, 'type': 'Goal', 'detail': 'Own Goal', 'side': 'home'},
      {'minute': 30, 'type': 'Goal', 'detail': 'Penalty', 'side': 'away'},
    ]);
    expect([for (final g in prog) '${g.home}-${g.away}'], [
      '1-0',
      '1-1',
      '1-2',
    ]);
    expect(prog[1].ownGoal, isTrue);
    expect(prog[1].side, 'away', reason: 'kendi kalesine gol rakibe yazılır');
    expect(prog[2].penalty, isTrue);
  });

  test('gol akışı: takımı belirsiz gol skora yazılmaz', () {
    expect(
      goalProgression([
        {'minute': 5, 'type': 'Goal', 'detail': 'Normal Goal'},
      ]),
      isEmpty,
    );
  });

  test('baskı: gerçek istatistiklerden pay hesaplanır, toplam 100', () {
    final pr = pressureIndex(_stats);
    expect(pr, isNotNull, reason: 'yeterli veri var');
    expect(pr!.home + pr.away, 100);
    expect(
      pr.home > 60,
      isTrue,
      reason: 'her ölçüde üstün takım baskıda görünür',
    );
    expect(pr.basis, contains('Şut'));
    expect(pr.basis, contains('Topla oynama'));
    expect(pr.basis, isNot(contains('Faul')), reason: 'faul baskı ölçüsü değil');
  });

  test('baskı: yetersiz veride GÖSTERGE ÜRETİLMEZ (uydurma yok)', () {
    expect(pressureIndex([]), isNull);
    expect(pressureIndex(null), isNull);
    expect(
      pressureIndex([
        {'type': 'Total Shots', 'home': 4, 'away': 2},
      ]),
      isNull,
      reason: 'tek ölçü yetmez',
    );
    expect(
      pressureIndex([
        {'type': 'Total Shots', 'home': null, 'away': 2},
        {'type': 'Corner Kicks', 'home': 3, 'away': null},
      ]),
      isNull,
      reason: 'eksik taraf sayılmaz',
    );
    expect(
      pressureIndex([
        {'type': 'Total Shots', 'home': 0, 'away': 0},
        {'type': 'Corner Kicks', 'home': 0, 'away': 0},
      ]),
      isNull,
      reason: '0-0 veri taşımaz',
    );
  });

  test('baskı: tek eksik ölçü diğerlerini bozmaz', () {
    final pr = pressureIndex([
      {'type': 'Total Shots', 'home': 10, 'away': 10},
      {'type': 'Corner Kicks', 'home': 5, 'away': 5},
      {'type': 'Shots on Goal', 'home': null, 'away': 3},
    ]);
    expect(pr!.home, 50);
    expect(pr.basis.length, 2);
  });

  test('istatistik sıralama: önemli satırlar üstte, bilinmeyen sona', () {
    final sorted = sortStats([
      {'type': 'Offsides'},
      {'type': 'Total Shots'},
      {'type': 'Bilinmeyen'},
      {'type': 'Ball Possession'},
    ]);
    expect([for (final s in sorted) s['type']], [
      'Ball Possession',
      'Total Shots',
      'Offsides',
      'Bilinmeyen',
    ]);
    expect(kRegMinutes, 90);
  });
}
